### PR TITLE
[TTS] FastPitch adapter fine-tune and conditional layer normalization

### PR DESCRIPTION
**Pull request by hsiehjackson**
# What does this PR do ?

Adds FastPitch pre-training with CLNs and fine-tuning with adapters.

**Collection**: TTS

# Changelog 
- Adds multi-speaker FastPitch pre-training with Conditional Layer Normalization 
    - `nemo/collections/tts/modules/fastpitch.py`
    - `nemo/collections/tts/modules/transformer.py`
    - `nemo/collections/tts/modules/submodules.py`
- Add adapter modules for FastPitch fine-tuning
    - `nemo/collections/tts/models/fastpitch.py`
    - `nemo/collections/tts/modules/fastpitch.py`
    - `nemo/collections/tts/modules/transformer.py`
    - `nemo/collections/tts/modules/aligner.py`
    - `nemo/collections/tts/parts/mixins/__init__.py`
    - `nemo/collections/tts/parts/mixins/fastpitch_adapter_mixins.py`
- Add config and fine-tuning python script
    - `examples/tts/conf/fastpitch_speaker_adaptation.yaml`
    - `examples/tts/fastpitch_finetune_adapters.py`
- Fix aligner `nan` loss bug
    - `nemo/collections/tts/losses/aligner_loss.py`
 
# Usage

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Comment by subhankar-ghosh** [comment]
LGTM. Will wait for others to review.

File: nemo/collections/tts/modules/transformer.py [hunk 1]
```diff
@@ -68,17 +70,17 @@ def __init__(self, d_model, d_inner, kernel_size, dropout, pre_lnorm=False):
             nn.Conv1d(d_inner, d_model, kernel_size[1], 1, (kernel_size[1] // 2)),
             nn.Dropout(dropout),
         )
-        self.layer_norm = nn.LayerNorm(d_model)
+        self.layer_norm = ConditionalLayerNorm(d_model, spk_emb_dim=d_model, condition=condition_lnorm)
```
Comment (rlangman) [inline nemo/collections/tts/modules/transformer.py:73]: If condition_lnorm=False then it should use the original nn.LayerNorm definition. Otherwise checkpoints will break. And I don't want to use a custom written LN over the torch defined layer one unless it is absolutely required.

Comment (hsiehjackson) [reply nemo/collections/tts/modules/transformer.py:73]: I think when we set condition_lnorm=False. `ConditionalLayerNorm` with the same as `torch.nn.LayerNorm`. We have the affine weight and bias and call the F.layer_norm. https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/normalization.py#L165-L195

Comment (hsiehjackson) [reply nemo/collections/tts/modules/transformer.py]: Fixed with inherit from `torch.nn.LayerNorm` to use original LN

File: nemo/collections/tts/modules/aligner.py [hunk 2]
```diff
@@ -175,3 +180,56 @@ def forward(self, queries, keys, mask=None, attn_prior=None, conditioning=None):
 
         attn = self.softmax(attn)  # softmax along T2
         return attn, attn_logprob
+
+
+class AlignmentEncoderAdapter(AlignmentEncoder, adapter_mixins.AdapterModuleMixin):
```
Comment (rlangman) [inline nemo/collections/tts/modules/aligner.py:185]: Do you think it would look more organized if we had an adapters.py with all available adapter modules?

**Review by rlangman — COMMENTED** [review]
As mentioned, it would be very useful to have a unit test that validates that the output of ConditionalLayerNorm(condition_types=[]) equals nn.LayerNorm. Though there is some overhead to add simply because we have no pre-existing unit test class for submodules.py.

Otherwise looks good to me, with some minor questions/comments.

File: nemo/collections/tts/modules/aligner.py [hunk 3]
```diff
@@ -151,13 +157,12 @@ def forward(self, queries, keys, mask=None, attn_prior=None, conditioning=None):
...
             attn_prior (torch.tensor): prior for attention matrix.
-            conditioning (torch.tensor): B x T2 x 1 conditioning embedding
+            conditioning (torch.tensor): B x 1 x C2 conditioning embedding
         Output:
             attn (torch.tensor): B x 1 x T1 x T2 attention mask. Final dim T2 should sum to 1.
             attn_logprob (torch.tensor): B x 1 x T1 x T2 log-prob attention mask.
         """
-        if conditioning is not None:
-            keys = keys + conditioning.transpose(1, 2)
+        keys = self.cond_input(keys.transpose(1, 2), conditioning)
```
[the remark this answers is not in the document]
Comment (hsiehjackson) [reply nemo/collections/tts/modules/aligner.py]: Yes. The output needs to be transposed back to B x C2 x T2. 
For `forward_time_minor()`, can you please explain more?

File: nemo/collections/tts/modules/submodules.py [hunk 4]
```diff
@@ -410,3 +416,92 @@ def forward(self, forward_input: Tuple[torch.Tensor, torch.Tensor]):
                 output = output + res_skip_acts
 
         return self.end(output)
+
+
+def check_support_condition_types(condition_types):
+    support_types = ["add", "concat", "layernorm"]
```
[the remark this answers is not in the document]
Comment (hsiehjackson) [reply nemo/collections/tts/modules/submodules.py:422]: You mean a constant global variable? 
Or maybe I create a `ConditionalInterface` to let `ConditionalInput` and `ConditionalLayerNorm` to inherit?

[the remark this answers is not in the document]
Comment (rlangman) [reply nemo/collections/tts/modules/submodules.py:422]: Yea, constant global variable

```
SUPPORTED_CONDITION_TYPES = ["add", "concat", "layernorm"]

def check_support_condition_types()
```

File: nemo/collections/tts/modules/aligner.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rlangman) [reply nemo/collections/tts/modules/aligner.py]: We have inconsistent conventions where most public methods are "time major" (B X T x C) and most layer definitions are "time minor" (B x C x T).

So we have lots of inline transposes like https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/tts/modules/fastpitch.py#L91

So was wondering if it would be worth the layers having a second method like:

```
def forward_time_minor(self, input):
    input = input.transpose(1, 2)
    output = self.forward(input)
    output = output.transpose(1, 2)
    return output
```

File: nemo/collections/tts/modules/aligner.py (hunk 2 above)
Comment (hsiehjackson) [reply nemo/collections/tts/modules/aligner.py:185]: Yes. We can do that. Then where should we import `adapters.py` since we should register the adapter class in a global way.

File: nemo/collections/tts/modules/submodules.py [hunk 5]
```diff
@@ -410,3 +416,92 @@ def forward(self, forward_input: Tuple[torch.Tensor, torch.Tensor]):
...
+
+        if "add" in self.condition_types and condition_dim != hidden_dim:
+            self.add_proj = torch.nn.Linear(condition_dim, hidden_dim)
+
+        if "concat" in self.condition_types:
+            self.concat_proj = torch.nn.Linear(hidden_dim + condition_dim, hidden_dim)
+
+    def forward(self, signal, conditioning=None):
+        """
+        signal: B x T x C
```
[the remark this answers is not in the document]
Comment (hsiehjackson) [reply nemo/collections/tts/modules/submodules.py]: Add the comment follow aligner.

File: nemo/collections/tts/modules/transformer.py [hunk 6]
```diff
@@ -157,29 +159,46 @@ def _forward(self, inp, attn_mask=None):
...
-        output = self.dec_attn(dec_inp, attn_mask=~mask.squeeze(2))
+    def forward(self, dec_inp, mask=None, conditioning=None):
+        output = self.dec_attn(dec_inp, attn_mask=~mask.squeeze(2), conditioning=conditioning)
         output *= mask
-        output = self.pos_ff(output)
+        output = self.pos_ff(output, conditioning)
         output *= mask
+
+        if self.is_adapter_available():
+            output = self.forward_enabled_adapters(output)
```
[the remark this answers is not in the document]
Comment (hsiehjackson) [reply nemo/collections/tts/modules/transformer.py:181]: Added

File: nemo/collections/tts/modules/aligner.py (hunk 2 above)
Comment (hsiehjackson) [reply nemo/collections/tts/modules/aligner.py]: Add `modules/adapters.py`.

File: nemo/collections/tts/modules/aligner.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (hsiehjackson) [reply nemo/collections/tts/modules/aligner.py]: I see. Thanks for the suggestion!

**Comment by hsiehjackson** [comment]
@rlangman I adjust following your comments and have included a unit test for the `ConditionalLayerNorm` module.

File: tests/collections/tts/modules/test_submodules.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+
+@pytest.mark.unit
+def test_conditional_layer_norm():
+    # NLP Example
+    batch, sentence_length, embedding_dim = 20, 5, 10
+    embedding = torch.randn(batch, sentence_length, embedding_dim)
+    lm = torch.nn.LayerNorm(embedding_dim)
+    clm = submodules.ConditionalLayerNorm(embedding_dim)
+    assert torch.all(lm(embedding) == clm(embedding))
```
Comment (rlangman) [inline tests/collections/tts/modules/test_submodules.py:28]: Lets randomize the weights to simulate the trained layers.

I think this works?

```
weight = torch.nn.Parameter(torch.randn(embedding_dim))
bias = torch.nn.Parameter(torch.randn(embedding_dim))

ln.weight = weight
ln.bias = bias
```

File: examples/tts/fastpitch_finetune_adapters.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
```
Comment (rlangman) [inline examples/tts/fastpitch_finetune_adapters.py:1]: Change to "NVIDIA CORPORATION & AFFILIATES." in every file

**Comment by subhankar-ghosh** [comment]
LGTM.

**Review by rlangman — COMMENTED** [review]
LGTM
